app/src/main/java/net/programmierecke/radiodroid2/players/exoplayer/IcyDataSource.java

```diff
@@ -209,33 +191,26 @@ public int read(byte[] buffer, int offset, int readLength) throws HttpDataSource
             return bytesTransferred;
         } catch (HttpDataSourceException readError) {
             dataSourceListener.onDataSourceConnectionLost();
+            throw readError;
+        }
+    }
 
-            final long reconnectStartTime = System.currentTimeMillis();
-
-            while (true) {
-                final long currentTime = System.currentTimeMillis();
-
-                if (BuildConfig.DEBUG) Log.d(TAG, "Reconnecting...");
-
-                try {
-                    reconnect();
-                    break;
-                } catch (HttpDataSourceException ex) {
-                    try {
-                        Thread.sleep(delayBetweenReconnections);
-                    } catch (InterruptedException e) {
-                        break;
-                    }
-                }
-
-                if (currentTime - reconnectStartTime > timeUntilStopReconnecting) {
-                    dataSourceListener.onDataSourceConnectionLostIrrecoverably();
-                    throw new HttpDataSourceException("Reconnection retry time ended.", dataSpec, HttpDataSourceException.TYPE_READ);
-                }
+    private void sendToDataSourceListenersWithoutMetadata(byte[] buffer, int offset, int bytesRead) {
+        if (bytesRead > remainingUntilMetadata) {
+            if (remainingUntilMetadata > 0) {
+                dataSourceListener.onDataSourceBytesRead(buffer, offset, remainingUntilMetadata);
             }
+            metadataBytesToSkip = buffer[offset + remainingUntilMetadata] * 16 + 1;
+            remainingUntilMetadata = shoutcastInfo.metadataOffset + remainingUntilMetadata + metadataBytesToSkip;
         }
 
-        return 0;
+        if (bytesRead > metadataBytesToSkip) {
+            dataSourceListener.onDataSourceBytesRead(buffer, offset +  metadataBytesToSkip, bytesRead - metadataBytesToSkip);
```

Comment:
I'm a bit confused with the skipping. So we have:
- `offset` - where we start reading
- `bytesRead` how much was read starting from `offset`
- `metadataBytesToSkip` - metadata size

So the check of `bytesRead > metadataBytesToSkip` and the read by offset to `offset +  metadataBytesToSkip` doesn't make sense to me, we can have metadata at the end of the buffer.

Reply:
If the metadata is at the end of the buffer, i.e. `bytesRead <= metadataBytesToSkip` the else part comes into play. The part of the audio before the metadata was already sent in line 201 and the rest ist skipped in the next call.

Reply:
`metadataBytesToSkip` is the size of metadata, the metadata itself could be at the end of the buffer, so `bytesRead` could be greater than `metadataBytesToSkip` while the end of metadata could be even past the read amount of bytes.

Reply:
Yes, there is a reading of buffer before metadata, but it doesn't change `bytesRead` or `offset`

Reply:
But the case, when `bytesRead > metadataBytesToSkip` is handled in the if part and if `bytesRead > metadataBytesToSkip`, then `bytesRead - metadataBytesToSkip > 0`.

I guess, I don't understand what you mean.

Reply:
E.g. you have 128 bytes of metadata, `bytesRead` is 2048, `remainingUntilMetadata` is 512, so you:
- read 512 bytes
- read `metadataBytesToSkip` which is 128
- check `bytesRead > metadataBytesToSkip` which is `2048 > 128` and do read something from `offset + 128` instead of from `offset + 512 + 128`.

Reply:
In such cases it's always better to draw several cases on paper and see how it goes.

Reply:
> E.g. you have 128 bytes of metadata, `bytesRead` is 2048, `remainingUntilMetadata` is 512, so you:
> 
> * read 512 bytes

I don't understand this value. If `bytesRead=2048`, where does the 512 come from?
This method is only for sending the audio bytes for recording listeners that cannot handle metadata.

What might be not easy to understand here is that here `metadataBytesToSkip` here is typically 0. I'm  handling here the case where `metadataBytesToSkip = 0` and `0 < metadataBytesToSkip < bytesRead` in one.

I was about to draw this case decision on paper, but got it right at the first attempt, so that that wasn't necessary. Well, it was the second attempt because initially I forgot to add the 1  for the byte that encodes the metadata frame length itself.

It's also working, perfectly, isn't it? Even when you stop your internet connections the recordings don't have any metadata residues and otherwise no gaps or garbled passages (I have recorded and checked 30 minutes of MANGORADIO :) and others, too) .

OTOH, I admit that this not really an argument for correctness. Particularly, because ExoPlayer seems to issue a single byte read for metadata frame length and afterwards a read for the length of the metadata frame length - which might change in future ExoPlayer versions.

Reply:
I agree that it works and that ExoPlayer has weird pattern of access. I think I should have started with picture:

![image](https://user-images.githubusercontent.com/3993671/80291124-831d6b00-8753-11ea-8c38-64c4691a3f9a.png)

Is it how you see it? 4) is ExoPlayer pattern.

Reply:
I don't if your visualisation in 4 is right. The constants i see in what ExoPlayer wants to read are are the following:

`if (remainingUntilMetadate == 0) readLength == 1`
`if (metadataBytesToSkip == x && x > 0) readLength == x`

(I'm not completely sure about the second one, but it looks like it.)

Reply:
4) May not be exactly right, but I'm trying to make a point that 1,2,3 are also possible in general case and we should not depend on exoplayer doing certain pattern of access.

Reply:
My code does not depend a specific case, I think.

Reply:
I only discovered this behaviour afterwards.

Reply:
Then, let's take the first case: how the second part of "Music" is being read? At which place to we do read starting  from `offset + remainingUntilMetadata + remainingUntilMetadata`?

Reply:
Hm, this function is not about reading. It'just about passing the read data without metadata to the the recorder. We read whatever ExoPlayer wants.

Reply:
ExoPlayer always gets the whole data. It's only about removing the metadata for the recording here.

Reply:
And how we would pass the second part of "Music" in the 1 case? Sorry for pressing here, but I just don't understand how callback would be called with right offset and length.

Reply:
The first part is passed in line 201 and the second part in line 208.

And if there is no metadata involved, everything is passed in 208.

Reply:
So instead of deleting the metadata, I simple pass the  part before the metadata and the part after the metadata separately .

Reply:
Line 208 passes `offset +  metadataBytesToSkip` which doesn't look like a start of the second "Music" piece. `metadataBytesToSkip` is just the size of metadata.

Reply:
Your case 3 is handled in 211.

Reply:
I understand the intent, however maths doesn't check out for me.

Reply:
However: I indeed do not handle the case when more than one metadata frame is in the read data. This could actually happen in hls streams. Maybe we can find some stream to test this.

Reply:
> more than one metadata frame is in the read data. 

Yes, it would be great if this also would be handled.

Reply:
> I understand the intent, however maths doesn't look right out for me.

Find a stream to prove me wrong, or let beta testers do the job. Anyway, problems will only become more likely when we switch to (very) future ExoPlayer versions. 
Anyway, I'll have another look tomorrow.

Reply:
Yes, it better to look with fresh head. Assign some values to variables in 1 case and walk line by line through code.

Reply:
> > more than one metadata frame is in the read data.
> 
> Yes, it would be great if this also would be handled.

Sounds like a good exercise for tomorrow. However, we will not be able to really test this, even  when your mock up streaming server for unit tests is ready.

Reply:
We can add unit tests =)

Reply:
> We can add unit tests =)

In addition we'd need to force ExoPlayer to swallow more read bytes than it wanted. I don't know if that's possible :)

Reply:
Now it looks correct, thank you!

Reply:
Thanks, it's indeed correct now :)

Should I add the corresponding unit tests in this PR?

Reply:
If you have them - yes.

Reply:
OK, done. However, I haven't succeeded in integrating the test in the gradle script.
It works with Android Studio.